keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,774 @@
+A `BucketAccess` instance is deleted only once the underlying Pod stop using the access credentials. `BucketAccess` instance is deleted by deleting the corresponding BAR or admin can delete the BA directly. If a Pod is using the BA then the BA deletion will wait until Pod is deleted. If a Pod using the BA is terminated, then the BA can be deleted by deleting the BAR. 
+
+> Note: it's up to each provisioner whether or not to physically delete bucket content. But, for idempotency, the expectation is that the backend bucket will at least be made unavailable, pending out-of-band bucket lifecycle policies.
+
+> Note: delete is described below as a synchronous workflow but it will likely need to be implemented asynchronously to accommodate potentially long delete times when buckets contain many objects. The steps should still mostly follow what's outlined below.
+
+Here is the workflow:
+
++ User deletes their `BucketRequest` (BR) which is deferred until finalizers have been removed.
++ COSI gets the `Bucket` instance from the BR and notes the retention policy.
++ If the `Bucket` instance was created by COSI then COSI deletes the `Bucket`, which sets its deleteTimestamp but does not delete it due to finalizer.
++ If the `Bucket` instance was not created by COSI then the `Bucket` is not deleted, but finalizers are removed so that the admin can manually delete it.
++ COSI unbinds the BR from the `Bucket`.
++ Sidecar sees the `Bucket` is unbound, and if the retention policy is "Delete", the gRPC calls the provisoner's _Delete_ interface. Upon successful completion, the `Bucket` is updated to Deleted.
++ If the retention policy is "Retain" then the `Bucket` remains "Released" and it can potentially be reused.
++ When COSI sees the `Bucket` is "Deleted", it removes all finalizers and the resources are garbage collected.
+
+## Grant Bucket Access
+This workflow describes the automation supporting granting access to an existing backend bucket.
+
+Here is the workflow:
+
++ The admin creates a cluster-scoped `Bucket` representing the backend bucket.
++ The user creates a `BucketRequest` referencing this `Bucket` instance and COSI follows the "Create Bucket" workflow.
++ The user creates a `BucketAccessRequest` (BAR) which references their BR.
++ COSI central controller sees a new `BucketAccessRequest` (BAR) and gets the `BAR.BucketAccessClass` (BAC).
++ COSI creates the cluster-scoped `BucketAccess` (BA) filling in the field from the BAR and BAC.
++ The sidecar sees the new BA and gRPC calls the driver to grant access to the backend bucket.
++ COSI updates the BAR and BA to reference each other.
++ The cosi-node-adapter, which has been waiting for the BAR to be processed, writes access tokens to the app pod's specified mount point.
+
+## Revoke Bucket Access
```

Comment:
In the last meeting we discussed Admin forcing access revocation by deleting the BA.
@jeffvance Should we add this flow?

Reply:
Yes. I think it's worth describing what we expect when the admin deletes the B or BA (assuming their finalizers are still in place).